test_cluster_detection.py

```diff
@@ -0,0 +1,242 @@
+
+        # Test
+        assert(not model_ran)
+        assert(detections_in_world is None)
+
+    def test_good_data(self, cluster_model:ClusterEstimation):
+        """
+        All conditions met should run
+        """
+        EXPECTED_CLUSTER_COUNT = MIN_TOTAL_POINTS_THRESHOLD + 1  # more than min total threshold should run 
+        generated_detections, labels, cluster_positions= generate_cluster_data([EXPECTED_CLUSTER_COUNT])
+    
+        # Run
+        cluster_model.clear_all_data()
+        model_ran, detections_in_world = cluster_model.run(generated_detections, False)
+        
+        # Test
+        assert(model_ran)
+        assert(detections_in_world is not None)
+    
+
+class TestCorrectNumberClusterOutputs():
+    """
+    Tests if cluster estimation output matches number of input clusters
+    at different data scenarios. 
+    """
+    def test_detect_normal_data(self, cluster_model:ClusterEstimation):
+        """
+        data with small distribution and equal number of points per cluster center. 
+        """
+        # Setup
+        STD_DEV_REGULAR = CENTER_BOX_SIZE / 500
+        EXPECTED_CLUSTER_COUNT = 5
+        DATA = [100, 100, 100, 100, 100]
+        generated_detections, labels, cluster_positions= generate_cluster_data(DATA, STD_DEV_REGULAR)
+
+        # Run
+        cluster_model.clear_all_data()
+        cluster_model.reset_model()
+        model_ran, detections_in_world = cluster_model.run(generated_detections, False)
+        
+        # Test
+        assert(model_ran)
+        assert(detections_in_world is not None)
+        assert(len(detections_in_world) == EXPECTED_CLUSTER_COUNT)
+
+    def test_detect_large_std_dev(self, cluster_model:ClusterEstimation):
+        """
+        data with large distribution and equal number of points per cluster center. 
+        """
+        # Setup
+        STD_DEV_LARGE = CENTER_BOX_SIZE / 100
+        EXPECTED_CLUSTER_COUNT = 5
+        DATA = [100, 100, 100, 100, 100]
+        generated_detections, labels, cluster_positions= generate_cluster_data(DATA, STD_DEV_LARGE)
+
+        # Run
+        cluster_model.clear_all_data()
+        cluster_model.reset_model()
+        model_ran, detections_in_world = cluster_model.run(generated_detections, False)
+        
+        # Test
+        assert(model_ran)
+        assert(detections_in_world is not None)
+        assert(len(detections_in_world) == EXPECTED_CLUSTER_COUNT)
+
+    def test_detect_skewed_data(self, cluster_model:ClusterEstimation):
+        """
+        data with small distribution but varying number of points per cluster center and
+        random outlier points to simulate false detections
+        """
+        # Setup
+        STD_DEV_REGULAR = CENTER_BOX_SIZE / 500
+        EXPECTED_CLUSTER_COUNT = 5
+        DATA = [1, 1, 1, 1, 20, 100, 100, 100, 100]
+        generated_detections, labels, cluster_positions = generate_cluster_data(DATA, STD_DEV_REGULAR)
```

Comment:
Are the single points guaranteed to be far away from the other clusters?

Reply:
Yes, I checked with random_state = 0 in another file and plotted the results, I can send the image in discord of the positionings of points & predicted clusters if necessary. There is a clear false cluster 6th false cluster that was filtered out according to the test.